File: typescript/src/nuc/PROPERTY.ts
import * as estree from "../lang/estree/estree";
import * as graph from "../graph";
import * as state from "../state";

import Identifier from "../lang/ast/Identifier";
import NODE from "./NODE";
import REFERENCE from "./REFERENCE";

class PROPERTY extends NODE {
  object!: { value: any; properties: Record<string, PROPERTY> };
  name!: Identifier;
  value!: {
    before: (scope: any) => void;
    run: (scope: any) => any;
    graph: (scope: any) => any;
  };

  before(scope: any): void {
    this.value.before(scope);
  }

  run(scope: any): { value: any } | undefined {
    const evaluation = this.value.run(scope);

    const ref = this.object.value;
    const target =
      ref instanceof REFERENCE
        ? graph.retrieve((ref as any).link)
        : this.object;

    const appended = estree.append(target.resolve().node, this.name.node);
    const variable = new Identifier(appended);

    if (!evaluation) {
      state.delete(scope, variable.toString());
      return;
    }

    const assigned = state.assign(scope, variable.toString(), evaluation);
    return { value: assigned };
  }

  graph(scope: any): any {
    const ref = this.object.value;
    const target =
      ref instanceof REFERENCE
        ? graph.retrieve((ref as any).link)
        : this.object;

    target.properties[this.name.toString()] = this;
    return this.value.graph(scope);
  }
}

export default PROPERTY;

File: python/src/nucleoid/nuc/PROPERTY.py
"""
PROPERTY runtime node class.
"""
from typing import Any, Dict, Optional
from .NODE import NODE


class PROPERTY(NODE):
    """Property assignment node."""

    def __init__(self) -> None:
        """Initialize a PROPERTY instance."""
        super().__init__()
        self.object: Any = None
        self.name: Any = None
        self.value: Any = None

    def before(self, scope: Any) -> None:
        """
        Execute before the main run.

        Args:
            scope: The execution scope
        """
        if self.value:
            self.value.before(scope)

    def run(self, scope: Any) -> Optional[Dict[str, Any]]:
        """
        Execute property assignment.

        Args:
            scope: The execution scope

        Returns:
            Dictionary with value or None
        """
        from .. import graph, state
        from ..lang.ast import __Identifier__
        from ..lang.estree.estree import append
        from .REFERENCE import REFERENCE

        evaluation = self.value.run(scope) if self.value else None

        # Determine target object
        ref = self.object.value if hasattr(self.object, 'value') else None
        if isinstance(ref, REFERENCE):
            target = graph.retrieve(ref.link)
        else:
            target = self.object

        # Build variable path
        appended = append(target.resolve().node, self.name.node)
        variable = __Identifier__(appended)

        if not evaluation:
            state.state_instance.delete(scope, str(variable))
            return None

        assigned = state.state_instance.assign(scope, str(variable), evaluation)
        return {"value": assigned}

    def graph(self, scope: Any) -> Any:
        """
        Build dependency graph.

        Args:
            scope: The execution scope

        Returns:
            Graph result from value
        """
        from .. import graph
        from .REFERENCE import REFERENCE

        # Determine target object
        ref = self.object.value if hasattr(self.object, 'value') else None
        if isinstance(ref, REFERENCE):
            target = graph.retrieve(ref.link)
        else:
            target = self.object

        target.properties[str(self.name)] = self
        return self.value.graph(scope) if self.value else None
